# Numpy
### Если задания покажутся сложными, рекомендую пройти краткий курс по основам numpy [ссылка1](https://stepik.org/lesson/241329/step/1?unit=213910), [ссылка2](https://stepik.org/lesson/16462/step/1?unit=4283)

In [1]:
import numpy as np
import random
import scipy.stats as sps
import cv2


### Задание 1
В первой задаче вам предлагается перемножить две квадратные матрицы двумя способами -- без использования пакета ***numpy*** и с ним.

In [ ]:
# Для генерации матриц используем фукнцию random -- она используется для генерации случайных объектов
# функция sample создает случайную выборку. В качестве аргумента ей передается кортеж (i,j),
# здесь i -- число строк, j -- число столбцов.
a = np.random.sample((100, 100))
b = np.random.sample((100, 100))
# выведите размерность (ранг) каждой матрицы с помощью функции ndim.
# Используйте функцию shape, что она вывела?
# ========
print("ранг a =",a.ndim, "\nранг b =", b.ndim)
print("размер матрицы a =",a.shape,"\nразмер матрицы b =",b.shape)
# ========
print(a)
print(b)

ранг a = 2 
ранг b = 2
размер матрицы a = (100, 100) 
размер матрицы b = (100, 100)
[[0.72148252 0.49902307 0.20770638 ... 0.33671942 0.95206702 0.74956423]
 [0.3847105  0.39000667 0.07158308 ... 0.44182065 0.55461692 0.79419047]
 [0.36977766 0.84522289 0.81375479 ... 0.80019727 0.130253   0.99085645]
 ...
 [0.66315786 0.9104901  0.55463796 ... 0.87461817 0.22265493 0.65234536]
 [0.00410864 0.23633411 0.57259221 ... 0.84618615 0.30014199 0.04427065]
 [0.45351774 0.98380724 0.3479158  ... 0.40664825 0.67261334 0.42839505]]
[[0.46799287 0.84753929 0.34129417 ... 0.18016286 0.00753726 0.18646619]
 [0.54544347 0.23151092 0.53006635 ... 0.13402511 0.47833718 0.90285323]
 [0.14104978 0.56869649 0.54596792 ... 0.37600195 0.76603188 0.01219478]
 ...
 [0.2945149  0.41465568 0.47280359 ... 0.71805369 0.04970104 0.90218469]
 [0.71091511 0.71988421 0.09386753 ... 0.17738705 0.17740098 0.22575648]
 [0.90574853 0.63744821 0.50149732 ... 0.08160222 0.87500279 0.43609434]]


Без использования numpy

In [ ]:
def mult(a, b):
    rows_a = len(a)
    cols_a = len(a[0])
    cols_b = len(b[0])

    result = [[0 for _ in range(cols_b)] for _ in range(rows_a)]

    for i in range(rows_a):
        for j in range(cols_b):
            for k in range(cols_a):
                result[i][j] += a[i][k] * b[k][j]
    return result

C использованием numpy

In [ ]:
def np_mult(a, b):
    return np.matmul(a, b)

In [ ]:
%%time
# засечем время работы функции без NumPy
M1 = mult(a, b)

CPU times: user 704 ms, sys: 0 ns, total: 704 ms
Wall time: 710 ms


In [ ]:
%%time
# засечем время работы функции с NumPy
M2 = np_mult(a, b)

CPU times: user 1.81 ms, sys: 1.05 ms, total: 2.86 ms
Wall time: 1.52 ms




---


Огромная разница времени выполнения - решение с numpy намного быстрее. *(С Numpy 1.5 мс. без Numpy - 710 мс.)*


---



In [ ]:
# проверим корректность
np.abs(np.array(M1) - M2).sum()

1.5266010677805752e-11

### Задание 2

Вам подаются на вход два вектора `a` и `b` в трехмерном пространстве. Заполните их случайными числами. Реализуйте их скалярное произведение с помощью  `NumPy` и без. Засеките время работы, как это показано в заданиях выше.

In [ ]:
a = np.random.sample((1, 3))
a = list(a)[0]
b = np.random.sample((1, 3))
b = list(b)[0]
print(a, b)

[0.66553995 0.81307984 0.1899588 ] [0.09681159 0.35949677 0.717094  ]


In [ ]:
def scalar_product(a, b):
    # Скалярным произведением двух векторов в трехмерном пространстве называется сумма произведений соответствующих координат векторов
    result = 0
    for i in range(3):
        result += a[i] * b[i]

    return result

In [ ]:
def np_scalar_product(a,b):
    return np.dot(a, b)

In [ ]:
%time product_1 = scalar_product(a,b)
%time product_2 = np_scalar_product(a,b)
# проверим корректность:
np.abs(product_1 - product_2).sum()

CPU times: user 34 µs, sys: 0 ns, total: 34 µs
Wall time: 40.1 µs
CPU times: user 66 µs, sys: 0 ns, total: 66 µs
Wall time: 69.9 µs


0.0

Почему методы ***numpy*** оказываются эффективнее? Что вы можете сказать о скорости вычислений в NumPy?

**Ответ:** Методы NumPy оказываются эффективнее, потому что они реализуются на низкоуровневых языках (**C, Fortran**). Кроме того, методы NumPy тщательно оптимизированы. Например, в документации к методу numpy.matmul написано:

> *It uses an optimized BLAS library when possible*

На сайте **superfastpython.com** про BLAS написано следующее:

> NumPy использует библиотеки BLAS и LAPACK для эффективного выполнения функций линейной алгебры с векторами и матрицами, что позволяет NumPy наилучшим образом использовать доступное системное оборудование.Современные реализации библиотеки BLAS, такие как OpenBLAS, позволяют NumPy эффективно и без проблем выполнять векторные и матричные операции, используя несколько процессорных ядер и наборы команд для конкретного процессора




------------------

### Задание 3
Напишите функцию, которая находим сумму четных диагональныех элементов квадратной матрицы. Если таких элементов нет, то вывести `0`.

In [ ]:
# функция, решающая задачу с помощью NumPy
def np_diag_2k(a):
    diag_elems = np.diagonal(a)
    even_diag = [i for i in diag_elems if i%2==0]

    if len(even_diag)>0:
        return np.sum(even_diag)
    else:
        return 0

In [ ]:
# зададим некоторую квадратную матрицу
a = np.random.randint(1, 10, size=(5, 5))
a

array([[7, 1, 4, 9, 7],
       [8, 8, 9, 6, 1],
       [5, 4, 8, 2, 7],
       [3, 5, 3, 3, 1],
       [2, 6, 5, 5, 1]])

In [ ]:
%%time
# засечем время работы функции с NumPy
np_diag_2k(a)

CPU times: user 151 µs, sys: 0 ns, total: 151 µs
Wall time: 159 µs


16

### Задание 4
На вход дан двумерный массив $X$. Напишите функцию, которая для каждой строчки $x = (x_1, x_2, \ldots, x_n)$ массива $X$  строит строчку $s = (s_1, s_2, \ldots, s_n)$, где $s_k=\frac{x_1+...+x_k}{k}$, а затем выдаёт массив из построенных строчек. Используйте библиотеку numpy (вам поможет функция cumsum). Выходом функции должен быть двумерный массив той же формы, что и $X$.

In [2]:
def np_sec_av(*A):
    #если выполняется условие, то массив двумерный
    if np.array(A).ndim-1>1:
        res = [[],[]]
        for i in A:
            row_cumsum = np.cumsum(i, axis=1)
            for j in range(len(i)):
                #print(j)
                for k in range(len(i[j])):
                    #проходимся по каждому элементу и делим его на КОЛИЧЕСТВО элементов, суммой которых он являтся
                    r = row_cumsum[j][k] / (k+1)
                    res[j].append(r)
    else:
        res = []
        row_cumsum = np.cumsum(A)
        for k in range(len(row_cumsum)):
            r = row_cumsum[k]/(k+1)
            res.append(r)

    return res

In [5]:
def check_dimension(arr):
    # Проверяем является ли массив двумерным
    if all(isinstance(sublist, list) for sublist in arr):
        if all(len(sublist) == len(arr[0]) for sublist in arr):
            return 2

    # Проверяем является ли массив одномерным
    if all(not isinstance(elem, list) for elem in arr):
        return 1

    return False

def get_cumsum(arr, dim):
    cumsum = []
    cum_sum = 0
    if dim == 2:
        for row in arr:
            cumsum_row = []
            for num in row:
                cum_sum += num
                cumsum_row.append(cum_sum)
            cumsum.append(cumsum_row)
        return cumsum
    else:
        cumsum_row = []
        for num in arr:
            cum_sum += num
            cumsum_row.append(cum_sum)
        return cumsum_row

#без numpy
def sec_av(A):
    if check_dimension(A) == 2:
        res = [[],[]]
        row_cumsum = get_cumsum(A, 2)
        for j in range(len(A)):
            for k in range(len(A[j])):
                r = row_cumsum[j][k] / (k+1)
                res[j].append(r)
    elif check_dimension(A) == 1:
        res = []
        row_cumsum = get_cumsum(A, 1)
        for k in range(len(row_cumsum)):
            r = row_cumsum[k]/(k+1)
            res.append(r)
    return res


In [37]:
# зададим некоторую последовательность и проверим ее на ваших функциях.
# функция должна работать ~ в 50 раз быстрее, чем без Numpy
A = sps.uniform.rvs(size=10**5)
#print(A)

%time S2 = np_sec_av(A)
%time S3 = sec_av(A)

CPU times: user 290 ms, sys: 18.9 ms, total: 309 ms
Wall time: 311 ms
CPU times: user 727 ms, sys: 24.4 ms, total: 752 ms
Wall time: 754 ms



---
Получилось, что функция без NumPy работает в среднем ***всего в 2-3 раза*** медленнее, чем с NumPy.


---



:A = sps.uniform.rvs(size=10**3) чем больше вводить size этой последовательности, тем ближе будут значения cumsum к 0.5. Например, при 10^7 значение в самом конце списка составило 0.4999...

---



### Задание 5
Написать функцию для кодирование массива (Run-length encoding). Все подряд повторения элементов функция сжимает в один элемент и считает количество повторений. Функция возвращает кортеж из двух векторов одинаковой длины. Первый содержит элементы, а второй — сколько раз их нужно повторить.

Пример: encode(np.array([1, 2, 2, 3, 3, 1, 1, 5, 5, 2, 3, 3])) = (np.array[1, 2, 3, 1, 5, 2, 3]), np.array[1, 2, 2, 2, 2, 1, 2])

In [41]:
# функция, решающая задачу с помощью NumPy
def np_encode(seq_array):
    elements = []
    repeats = []
    change = (seq_array[1:] != seq_array[:-1])
    x = np.append(np.where(change), len(seq_array) - 1)
    counter = np.diff(np.append(-1, x))
    for i in range(0, len(counter)):
        elements.append(seq_array[x][i])
        repeats.append(counter[i])
    return (elements, repeats)

In [42]:
#функция решающая проблему без NumPy
def encode(seq):
  elements = []
  repeats = []
  count = 1
  char = seq[0]
  for i in range(1,len(seq)):
    if seq[i] == char:
        count = count + 1
    else :
        elements.append(char)
        repeats.append(count)
        char = seq[i]
        count = 1
  elements.append(char)
  repeats.append(count)
  return (elements,repeats)

In [54]:
from itertools import groupby
X = np.array([1, 2, 2, 3, 3, 1, 1, 5, 5, 2, 3, 3])
%time x, num = np_encode(X)
%time x, num = encode(X)

CPU times: user 860 µs, sys: 0 ns, total: 860 µs
Wall time: 847 µs
CPU times: user 23 µs, sys: 0 ns, total: 23 µs
Wall time: 25.5 µs


Огромный прирост в скорости у решения с NumPy - он в 10 и более раз быстрее решения без NumPy.

---



### Задание 6

Пусть задан некоторый одномерный массив $X$. Надо построить новый массив, где все элементы с нечетными индексами требуется заменить на число $a$ (если оно не указано, то на 1). Все элементы с четными индексами нужно возвести в куб. Затем записать элементы в обратном порядке относительно их позиций. Массив $X$ при этом должен остаться *без изменений*. В конце требуется слить массив X с преобразованным X и вывести в обратном порядке.  

Пример:  
X = [100, 200, 300, 400, 500] -> [100, a, 300, a, 500] -> [500^3, a, 300^3, a, 100^3] -> склеиваем -> [100, 200, 300, 400, 500, 500^3, a, 300^3, a, 100^3]

In [ ]:
# функция, решающая задачу с помощью NumPy
from copy import copy

def np_transformation(X, a=1):
    """
    param X: array
    """
    result = None
    pass
    return result

In [ ]:
X = np.array([i for i in range(1, 10, 2)])
# функция должна работать ~ в 20 раз быстрее, чем без Numpy
%time S2 = np_transformation(X, 5)